src/hooks/core/chat/useSystemMessage.ts

```diff
@@ -0,0 +1,160 @@
+    if (message?.text?.toLocaleLowerCase()?.includes('this topic has new')) return false;
+
+    console.log('checkpoint3');
+    // If the message is a system message, save the message
+    if (__isMessageOnlyForSystemUser(message) && __isMySystemMessage(message)) {
+      const newMessage = checkSystemMessageOnlyForSystemUser(message);
+      saveSystemMessageCallback?.(newMessage);
+      return true;
+    }
+
+    console.log('checkpoint 3.5');
+    if (__isMessageOnlyForSystemUser(message) && !__isMySystemMessage(message)) return false;
+
+    console.log('checkpoint4');
+    if (__isMessageForOtherUser(message)) {
+      const newMessage = checkSystemMessageOnlyForOtherSystemUser(message);
+      saveSystemMessageCallback?.(newMessage);
+      return true;
+    }
+
+    console.log('checkpoint5');
+    // If the message is a system message, save the message
+    if (__isSystemMessage(message)) {
+      const newMessage = checkSystemMessage(message);
+      saveSystemMessageCallback?.(newMessage);
+      return true;
+    }
+
+    console.log('checkpoint6');
+    if (message?.better_type) {
+      saveSystemMessageCallback?.(message);
+      return true;
+    }
+
+    console.log('checkpoint7');
+    return false;
+  }
+
+  function getFirstMessage(messages: GetstreamMessage[]): GetstreamMessage | null | undefined {
+    if (!messages) return null;
+    if (messages.length === 0) return null;
+    const sortedMessages = [...messages]?.sort((a, b) =>
+      moment(b?.updated_at).diff(moment(a?.updated_at))
+    );
+
+    let firstMessage: GetstreamMessage | null | undefined = null;
+    while (sortedMessages?.length > 0) {
+      const checkFirstMessage = sortedMessages?.at(0);
+      if (!__isSystemMessage(checkFirstMessage)) {
+        firstMessage = checkFirstMessage;
+        break;
+      }
+
+      if (checkFirstMessage?.text?.toLocaleLowerCase()?.includes('this topic has new')) {
+        sortedMessages.shift();
+        continue;
+      }
+
+      if (checkFirstMessage?.text?.toLocaleLowerCase()?.includes('you joinedt this community')) {
+        sortedMessages?.shift();
+        continue;
+      }
+
+      if (
+        __isMessageOnlyForSystemUser(sortedMessages[0]) &&
+        __isMySystemMessage(sortedMessages[0])
+      ) {
+        firstMessage = checkSystemMessageOnlyForSystemUser(sortedMessages[0]);
+        break;
+      }
+
+      if (__isSystemMessage(sortedMessages[0])) {
+        firstMessage = checkSystemMessage(sortedMessages[0]);
+        break;
+      }
+
+      if (checkFirstMessage?.text?.toLocaleLowerCase()?.includes('started following')) {
+        firstMessage = checkFirstMessage;
+        break;
+      }
+
+      sortedMessages.shift();
+    }
+
+    return firstMessage;
+  }
+
+  return {
+    saveSystemMessageFromWebsocket,
+    getFirstMessage
+  };
+};
+
+export default useSystemMessage;
+
+export type SaveSystemMessageCallback = ((message: GetstreamMessage) => Promise<void>) | undefined;
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The `useSystemMessage` hook seems to be well implemented. However, there are a few points that could be improved:

1. The use of console.log for debugging purposes is not recommended in production code. It's better to use a logging library that can be turned on or off based on the environment.

2. The function `saveSystemMessageFromWebsocket` has a high cyclomatic complexity due to multiple if conditions. Consider refactoring this function to reduce complexity and improve readability.

3. In the `getFirstMessage` function, you're sorting the messages array every time the function is called. If the messages array is large, this could be a performance issue. Consider sorting the messages once when they are received and then just picking the first one when needed.

4. The `getFirstMessage` function also uses a while loop with shift operation which can be costly as it re-indexes the array each time. Consider using a for loop instead.

5. There are several hardcoded strings used for comparison in the `getFirstMessage` function. These should ideally be moved to constants.

6. The `getFirstMessage` function seems to have some duplicated logic with `saveSystemMessageFromWebsocket`. Consider creating helper functions to avoid duplicating logic.

Here are some suggested changes:

```diff
- 51:     console.log('message', message);
- 60:     console.log('checkpoint1');
- 63:     console.log('checkpoint2');
- 67:     console.log('checkpoint3');
- 75:     console.log('checkpoint 3.5');
- 78:     console.log('checkpoint4');
- 85:     console.log('checkpoint5');
- 93:     console.log('checkpoint6');
- 99:     console.log('checkpoint7');

+ // Remove all console.log statements

- 106:     const sortedMessages = [...messages]?.sort((a, b) =>
- 107:       moment(b?.updated_at).diff(moment(a?.updated_at))
- 108:     );

+ // Sort messages once when they are received

- 111:     while (sortedMessages?.length > 0) {
- 119:         sortedMessages.shift();
- 124:         sortedMessages?.shift();
- 146:       sortedMessages.shift();

+ // Replace while loop with a for loop

- 65:     if (message?.text?.toLocaleLowerCase()?.includes('this topic has new')) return false;
- 118:       if (checkFirstMessage?.text?.toLocaleLowerCase()?.includes('this topic has new')) {
- 123:       if (checkFirstMessage?.text?.toLocaleLowerCase()?.includes('you joinedt this community')) {
- 141:       if (checkFirstMessage?.text?.toLocaleLowerCase()?.includes('started following')) {

+ // Move hardcoded strings to constants
```

Overall, the code is well structured and follows good practices. The use of TypeScript provides type safety and the use of hooks makes the code modular and reusable. However, there are some areas where performance could be improved and complexity reduced.


<!-- This is an auto-generated comment by OSS CodeRabbit -->